assets/js/components/GoogleChartV2.js

```diff
@@ -0,0 +1,174 @@
+/**
+ * GoogleChartV2 component.
+ *
+ * Site Kit by Google, Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+import { Chart } from 'react-google-charts';
+
+/**
+ * WordPress dependencies
+ */
+import { Fragment, useEffect, useLayoutEffect, useRef } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import PreviewBlock from './PreviewBlock';
+
+export default function GoogleChartV2( props ) {
+	const {
+		chartEvents,
+		chartType,
+		children,
+		className,
+		getChartWrapper,
+		height,
+		loaded,
+		loadingHeight,
+		loadingWidth,
+		onMouseOver,
+		onMouseOut,
+		onReady,
+		onSelect,
+		width,
+		...otherProps
+	} = props;
+
+	const loadingHeightToUse = loadingHeight || height;
+	const loadingWidthToUse = loadingWidth || width;
+	const loadingShape = chartType === 'PieChart' ? 'circular' : 'square';
+
+	const loader = (
+		<div className="googlesitekit-chart-v2-loading">
+			<PreviewBlock
+				className="googlesitekit-chart-v2-loading__wrapper"
+				height={ loadingHeightToUse }
+				shape={ loadingShape }
+				width={ loadingWidthToUse }
+			/>
+		</div>
+	);
+
+	const chartWrapperRef = useRef();
+	const googleRef = useRef();
+
+	useEffect( () => {
+		// Remove all event listeners after the component has unmounted.
+		return () => {
+			// eslint-disable-next-line no-unused-expressions
+			googleRef.current?.visualization.events.removeAllListeners( chartWrapperRef.current?.getChart() );
```

Comment:
@tofumatt we also need to remove all events assigned to the `chartWrapper` itself because this is how the library assigns `select` and `ready` events.
```suggestion
			if ( googleRef.current && chartWrapperRef.current ) {
				const { events } = googleRef.current.visualization;
				events.removeAllListeners( chartWrapperRef.current.getChart() );
				events.removeAllListeners( chartWrapperRef.current );
			}
```

Also, will we have the correct references to the google and chartWrapper instances here? Will useEffect memoize it and work with the original values? :thinking:

Reply:
These are the "previous" instances since we haven't yet re-assigned the `ref`. These refs should still be intact at this point, yeah.